Create mpas section (#136)
open-component-model/ocm-website: content/en/mpas/getting_started.md

```
We should also have a command to create a component subscription as well as a product deployment generator.
--->

1. Create a secret containing your Github credentials that will be used by MPAS to
create your project repository.

```bash
kubectl create secret generic \
  github-access \
  --from-literal=username=$GITHUB_USER \
  --from-literal=password=$GITHUB_TOKEN \
  -n mpas-system
```

2. Create a project that will contain the podinfo application.

Let's create a directory for the project:

```bash
mkdir -p ./clusters/my-cluster/podinfo
````

Then, create a `project.yaml` file in the `./clusters/my-cluster/podinfo` directory:

```bash
cat <<EOF >> ./clusters/my-cluster/podinfo/project.yaml
apiVersion: mpas.ocm.software/v1alpha1
kind: Project
metadata:
  name: podinfo-application
  namespace: mpas-system
spec:
  flux:
    interval: 1h
  git:
    provider: github
    owner: $GITHUB_USER
    isOrganization: false
    visibility: public
    maintainers:
    - $GITHUB_USER
    existingRepositoryPolicy: adopt
    defaultBranch: main
    credentials:
      secretRef:
        name: github-access
    commitTemplate:

>       email: mpas@mpas.ocm.software

Reviewer: This, should be `$GITHUB_EMAIL` I think considering this is something the user will have to make work. Potentially will be rejected if the token doesn't match with the email or something like that.

      message: Initializing Project repository
      name: mpas-admin
  prune: true
EOF
```

Then, apply the project to the clustern in a gitOps fashion:

```bash
git add --all && git commit -m "Add podinfo project" && git push
```

`Flux` will detect the changes and apply the project to the cluster.

This will create in the cluster a `namespace` for the project, a `service account`, and RBAC.
It will also create a GitHub repository for the project, and configure `Flux` to manage the project's resources.

3. Add the needed secrets to the namespace

> `Flux` is used to deploy all workloads in a gitOps way. In order for flux to access the internal

Reviewer: You're using `Flux` and flux and Flux interchangeably. I think it might be better if you stick to a format?

registry, we have to provide the certificate to use `https`.

```bash
kubectl get secret ocm-registry-tls-certs --namespace=mpas-system -o yaml | sed 's/namespace: 
.*/namespace: mpas-podinfo-application/' | kubectl apply -f -
```

We also need a secret in the project namespace that will be used to communicate with github:

```bash
kubectl create secret generic \
  github-access \
  --from-literal=username=$GITHUB_USER \
  --from-literal=password=$GITHUB_TOKEN \
  -n mpas-podinfo-application
```

> **Note** The credentials shall have access to github packages.

Reviewer: ```suggestion
**Note** The credentials should have access to GitHub packages.
```

As part of step 2, a `service account` was created for the project. We will use this service account
to provide the necessary permissions to pull from the `ghcr` registry.

> First create a secret containing the credentials for the service account:

Reviewer: ```suggestion
First, create a secret containing the credentials for the service account:
```

  --docker-email=<MY_EMAIL> -n mpas-podinfo-application
```

Then, patch the service account to use the secret:

```bash
kubectl patch serviceaccount mpas-podinfo-application -p '{"imagePullSecrets": [{"name": "github-registry-key"}]}' \
  -n mpas-podinfo-application
```

4. Clone the project repository

```bash
git clone https://github.com/$GITHUB_USER/mpas-podinfo-application
cd mpas-podinfo-application
```

5. Add the podinfo component subscription

Create a file under `./subscriptions/` that will contains the subscription declaration.

```bash
cat <<EOF >> ./subscriptions/podinfo.yaml
apiVersion: delivery.ocm.software/v1alpha1
kind: ComponentSubscription
metadata:
  name: podinfo-subscription
  namespace: mpas-podinfo-application
spec:
  interval: 30s
  component: mpas.ocm.software/podinfo
  semver: ">=v1.0.0"
  source:
    url: ghcr.io/open-component-model/mpas
    secretRef:
      name: github-access
  destination:
    url: ghcr.io/$GITHUB_USER
    secretRef:
      name: github-access
EOF
```

Then, apply the `ComponentSubscription` to the project in a gitOps fashion:

```bash
git add --all && git commit -m "Add podinfo subscription" && git push
```

> `Flux` will detect the changes and apply the subscription to the cluster.

Reviewer: Perhaps add a way to "kick" flux into action so people don't have to wait?

```
flux reconcile source git flux-system
```

Author: Yeah, but that means stating `flux cli` is required. It will make things easier but it's not mandatory. I did not want to have a long list of requirement.

This will replicate the product referenced by the `ComponentSubscription` `spec.component` field from
defined registry in the `spec.source.url` to the `spec.destination.url` registry.

6. Add a target for the podinfo application

The target will define where the application will be installed

<!-- TODO: 
The target must create the namespace with the desired service account.
We should leverage ESO to make sure that the needed imagePullSecret is attached to it.
--->

```bash
cat <<EOF >> ./targets/podinfo.yaml
apiVersion: mpas.ocm.software/v1alpha1
kind: Target
metadata:
  name: podinfo-kubernetes-target
  namespace: mpas-podinfo-application
  labels:

>     target.mpas.ocm.software/ingress-enabled: "true" # DO NOT CHANGE. expected by the component for target selection.

Reviewer: ```suggestion
    target.mpas.ocm.software/ingress-enabled: "true" # This label is defined by the component that will use it to select an appropriate target to deploy to.
```

```bash
git add --all && git commit -m "Add a target for podinfo" && git push
```

`Flux` will detect the changes and apply the target to the cluster.

7. Deploy the podinfo application

In order to deploy the podinfo application, we need to create a `ProductDeploymentGenerator` resource:

```bash
cat <<EOF >> ./generators/podinfo.yaml
apiVersion: mpas.ocm.software/v1alpha1
kind: ProductDeploymentGenerator
metadata:
  name: podinfo
  namespace: mpas-podinfo-application
spec:
  interval: 1m
  serviceAccountName: mpas-podinfo-application
  subscriptionRef:
    name: podinfo-subscription
    namespace: mpas-podinfo-application
EOF
```

Then, apply the `ProductDeploymentGenerator` to the project in a gitOps fashion:

```bash
git add --all && git commit -m "Add podinfo deployment generator" && git push
```

`Flux` will detect the changes and apply the resource to the cluster.

This will create a pull request in the project repository with the `ProductDeployment` resource
that will deploy the podinfo application.

<!--TODO: The pull request commit needs changed to something more meaningful--->

Go to the project repository and retrieve the pull request. 

> It should contains a `ProductDeployment` declaration that provides the configuration and

Reviewer: ```suggestion
It should contain a `ProductDeployment` declaration that provides the configuration and
```

all steps needed to deploy the product, as well as a `values.yaml` file. The `values` file

> contains values that should used to configure the different resources that are part of

Reviewer: ```suggestion
contains values that should be used to configure the different resources that are part of
```

the product to be deployed. There is a check that should pass before merging the pull request.

Once the pull request is merged, `Flux` will detect the changes and deploy the application to the cluster.

After a moment the `ProductDeployment` should be deployed successfully.
It is possible to verify with the command:

```bash
k describe productDeployment -n mpas-podinfo-application  
```

The result should something like:

```bash
Name:         podinfo
Namespace:    mpas-podinfo-application
Labels:       kustomize.toolkit.fluxcd.io/name=mpas-podinfo-application-products
              kustomize.toolkit.fluxcd.io/namespace=mpas-system
API Version:  mpas.ocm.software/v1alpha1
Kind:         ProductDeployment
Metadata:
...
Status:
  Conditions:
    Last Transition Time:  2023-09-14T10:14:41Z
    Message:               Reconciliation success
    Observed Generation:   1
    Reason:                Succeeded
    Status:                True
    Type:                  Ready
  Observed Generation:     1
```

> You should then have the application deployed in the `podinfo` namespace.

Reviewer: Hm, these shoulds read a bit differently. Maybe be a bit more assertive?

Is deployed, contains, etc...? That would read more decisive. WDYT?

--->

1. Create a secret containing your GitHub credentials that will be used by MPAS to
create your project repository.

```bash
kubectl create secret generic \
  github-access \
  --from-literal=username=$GITHUB_USER \
  --from-literal=password=$GITHUB_TOKEN \
  -n mpas-system
```

2. Create a project that will contain the podinfo application.

Let's create a directory for the project:

```bash
mkdir -p ./clusters/my-cluster/podinfo
````

Then, create a `project.yaml` file in the `./clusters/my-cluster/podinfo` directory:

```bash
cat <<EOF >> ./clusters/my-cluster/podinfo/project.yaml
apiVersion: mpas.ocm.software/v1alpha1
kind: Project
metadata:
  name: podinfo-application
  namespace: mpas-system
spec:
  flux:
    interval: 1h
  git:
    provider: github
    owner: $GITHUB_USER
    isOrganization: false
    visibility: public
    maintainers:
    - $GITHUB_USER
    existingRepositoryPolicy: adopt
    defaultBranch: main
    credentials:
      secretRef:
        name: github-access
    commitTemplate:
      email: <MY_EMAIL>
      message: Initializing Project repository
      name: mpas-admin
  prune: true
EOF
```

> Then, apply the project to the clustern in a gitOps fashion:

Reviewer 2: ```suggestion
Then, apply the project to the cluster in a gitOps fashion:
```

spec:
  type: kubernetes
  access:
    targetNamespace: podinfo
EOF
```

Then, apply the `Target` to the project in a gitOps fashion:

```bash
git add --all && git commit -m "Add a target for podinfo" && git push
```

`Flux` will detect the changes and apply the target to the cluster.

7. Deploy the podinfo application

In order to deploy the podinfo application, we need to create a `ProductDeploymentGenerator` resource:

```bash
cat <<EOF >> ./generators/podinfo.yaml
apiVersion: mpas.ocm.software/v1alpha1
kind: ProductDeploymentGenerator
metadata:
  name: podinfo
  namespace: mpas-podinfo-application
spec:
  interval: 1m
  serviceAccountName: mpas-podinfo-application
  subscriptionRef:
    name: podinfo-subscription
    namespace: mpas-podinfo-application
EOF
```

Then, apply the `ProductDeploymentGenerator` to the project in a gitOps fashion:

```bash
git add --all && git commit -m "Add podinfo deployment generator" && git push
```

`Flux` will detect the changes and apply the resource to the cluster.

This will create a pull request in the project repository with the `ProductDeployment` resource
that will deploy the podinfo application.

> <!--TODO: The pull request commit needs changed to something more meaningful--->

Reviewer 2: ```suggestion
<!--TODO: The pull request commit needs changed to something more meaningful-; should be configurable from project CR...-->
```
